### PR TITLE
TYP: Fix falsely rejected value types in ``ndarray.__setitem__``

### LINKED ISSUE #27964: TYP: False positives in ``ndarray.__setitem__`` with ``object_`` dtype in NumPy 2.2
### Describe the issue:

Under NumPy 2.2, assignment of a slice to an object array results in a type check failure

### Reproduce the code example:

```python
import numpy as np
x = np.zeros(1, dtype=object)
x[0] = slice(None)
```

### Error message:

```shell
tmp.py:3: error: No overload variant of "__setitem__" of "ndarray" matches argument types "int", "slice"  [call-overload]
tmp.py:3: note: Possible overload variants:
tmp.py:3: note:     def __setitem__(self, str | list[str], Buffer | _SupportsArray[dtype[Any]] | _NestedSequence[_SupportsArray[dtype[Any]]] | bool | int | float | complex | str | bytes | _NestedSequence[bool | int | float | complex | str | bytes], /) -> None
tmp.py:3: note:     def __setitem__(self, SupportsIndex | slice | EllipsisType | _SupportsArray[dtype[numpy.bool[builtins.bool]] | dtype[integer[Any]]] | _NestedSequence[_SupportsArray[dtype[numpy.bool[builtins.bool]] | dtype[integer[Any]]]] | builtins.bool | int | _NestedSequence[builtins.bool | int] | tuple[SupportsIndex | slice | EllipsisType | _SupportsArray[dtype[numpy.bool[builtins.bool]] | dtype[integer[Any]]] | _NestedSequence[_SupportsArray[dtype[numpy.bool[builtins.bool]] | dtype[integer[Any]]]] | builtins.bool | int | _NestedSequence[builtins.bool | int] | None, ...] | None, Buffer | _SupportsArray[dtype[Any]] | _NestedSequence[_SupportsArray[dtype[Any]]] | bool | int | float | complex | str | bytes | _NestedSequence[bool | int | float | complex | str | bytes], /) -> None
Found 1 error in 1 file (checked 1 source file)
```

### Python and NumPy Versions:

```
>>> import sys, numpy; print(numpy.__version__); print(sys.version)
2.2.0
3.12.3 (v3.12.3:f6650f9ad7, Apr  9 2024, 08:18:47) [Clang 13.0.0 (clang-1300.0.29.30)]
```

### Runtime Environment:

```python
>>> import numpy; numpy.show_runtime()
[{'numpy_version': '2.2.0',
  'python': '3.12.3 (v3.12.3:f6650f9ad7, Apr  9 2024, 08:18:47) [Clang 13.0.0 '
            '(clang-1300.0.29.30)]',
  'uname': uname_result(system='Darwin', node='jmdg-macbookpro.roam.internal', release='23.6.0', version='Darwin Kernel Version 23.6.0: Thu Sep 12 23:35:29 PDT 2024; root:xnu-10063.141.1.701.1~1/RELEASE_ARM64_T6000', machine='arm64')},
 {'simd_extensions': {'baseline': ['NEON', 'NEON_FP16', 'NEON_VFPV4', 'ASIMD'],
                      'found': ['ASIMDHP'],
                      'not_found': ['ASIMDFHM']}}]
```

### Context for the issue:

Related to #27957; this is a reduction of an error we're seeing in the JAX CI under NumPy 2.2

### PR DESCRIPTION
**Pull request by jorenham**
fixes https://github.com/numpy/numpy/issues/27964

cc @jakevdp

### CODE REVIEW

File: numpy/__init__.pyi [hunk 1]
```diff
@@ -982,6 +983,8 @@ if sys.version_info >= (3, 11):
     _ConvertibleToComplex: TypeAlias = SupportsComplex | SupportsFloat | SupportsIndex | _CharLike_co
 else:
     _ConvertibleToComplex: TypeAlias = complex | SupportsComplex | SupportsFloat | SupportsIndex | _CharLike_co
+_ConvertibleToTD64: TypeAlias = dt.timedelta | int | _CharLike_co | character | number | timedelta64 | np.bool | None
+_ConvertibleToDT64: TypeAlias = dt.date | int | _CharLike_co | character | number | datetime64 | np.bool | None
```
Comment (seberg) [inline numpy/__init__.pyi:987]: How come this has `character` and `_CharLike_co`?  Is that intentional?

Is this missing `dt.datetime`?

**Review by seberg — COMMENTED** [review]
This change looks fine to me (and introducing the `_ToIndices` seems nice).
But, I am not quite sure about the datetime definition.

I can't say whether it is all the ideal definitions, so if anyone else deeper into typing wants to have a look that would be nice.

**Comment by jorenham** [comment]
> I can't say whether it is all the ideal definitions, so if anyone else deeper into typing wants to have a look that would be nice.

There's a good chance I'm missing something here, so more eyes could indeed help in this case

File: numpy/__init__.pyi (hunk 1 above)
Comment (seberg) [reply numpy/__init__.pyi:987]: > I could sneak it in this PR, or open another PR for it?

Not sure I care, if you think it's simple enough (IIUC, this is about the scalar creation?).  FWIW, I would expect them to use the identical definitions to scalar assignment.

W.r.t to the strings, what surprises me was that `_ConvertibleToComplex` does not include `character`.

The other thing is that you have `_ConvertibleToDT64: TypeAlias = dt.date`.  Should it not also have `dt.datetime`?

**Review by seberg — APPROVED** [review]

**Comment by charris** [comment]
Let's get this in. I doubt it will be the last :)

Thanks Joren.

**Comment by jakevdp** [comment]
Thanks for the fix!

**Comment by jorenham** [comment]
> There's a good chance I'm missing something here

I missed something...

... and it's now causing the scipy CI to fail 😅

**Pull request merged**